### ISSUE TITLE
Undefined symbols for "typeinfo for __int128" when using Clang++ and libc++ on macOS

### ISSUE THREAD

**Issue by ArseniyKholod** [llvm/llvm-project#119608]
I have encountered a linker issue when compiling code with Clang and libc++ on macOS that involves `typeid(__int128)`. The compilation succeeds, but linking fails with the following error: `Undefined symbols for architecture arm64: "typeinfo for __int128" `. The same happens on x86_64.

It appears that `__int128` itself is supported as a type, but the runtime type information (`typeid`) for it is not being resolved by the linker.

Minimal working example (MWE):
test_clang.cpp:
```
#include <iostream>
#include <string>
#include <typeinfo>

int main()
{
  __int128 t = 1234567890;

  std::cout << typeid(__int128).name();

  return 0;
}
```

Compilation command: `/opt/homebrew/opt/llvm/bin/clang++ test_clang.cpp -stdlib=libc++ -std=gnu++17 -frtti -v`

If the `typeid` line is commented out, the code compiles and links without issues.

Is there a workaround to use `typeid` with `__int128`?
Will support for typeinfo on `__int128` be implemented in future versions of Clang/libc++?

The full error message:
```
bash-3.2$ /opt/homebrew/opt/llvm/bin/clang++ test_clang.cpp -stdlib=libc++ -std=gnu++17 -frtti -v
Homebrew clang version 19.1.4
Target: arm64-apple-darwin23.6.0
Thread model: posix
InstalledDir: /opt/homebrew/Cellar/llvm/19.1.4/bin
Configuration file: /opt/homebrew/etc/clang/arm64-apple-darwin23.cfg
System configuration file directory: /opt/homebrew/etc/clang
User configuration file directory: /Users/runner/.config/clang
 "/opt/homebrew/Cellar/llvm/19.1.4/bin/clang-19" -cc1 -triple arm64-apple-macosx14.0.0 -Wundef-prefix=TARGET_OS_ -Werror=undef-prefix -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name test_clang.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=non-leaf -ffp-contract=on -fno-rounding-math -funwind-tables=1 -fcompatibility-qualified-id-block-type-checking -fvisibility-inlines-hidden-static-local-var -fbuiltin-headers-in-system-modules -fdefine-target-os-macros -target-cpu apple-m1 -target-feature +zcm -target-feature +zcz -target-feature +v8.4a -target-feature +aes -target-feature +altnzcv -target-feature +ccdp -target-feature +complxnum -target-feature +crc -target-feature +dotprod -target-feature +fp-armv8 -target-feature +fp16fml -target-feature +fptoint -target-feature +fullfp16 -target-feature +jsconv -target-feature +lse -target-feature +neon -target-feature +pauth -target-feature +perfmon -target-feature +predres -target-feature +ras -target-feature +rcpc -target-feature +rdm -target-feature +sb -target-feature +sha2 -target-feature +sha3 -target-feature +specrestrict -target-feature +ssbs -target-abi darwinpcs -debugger-tuning=lldb -fdebug-compilation-dir=/Users/runner/work/openfhe-julia/openfhe-julia/test -target-linker-version 1115.7.3 -v -fcoverage-compilation-dir=/Users/runner/work/openfhe-julia/openfhe-julia/test -resource-dir /opt/homebrew/Cellar/llvm/19.1.4/lib/clang/19 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -stdlib=libc++ -internal-isystem /opt/homebrew/Cellar/llvm/19.1.4/bin/../include/c++/v1 -internal-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/local/include -internal-isystem /opt/homebrew/Cellar/llvm/19.1.4/lib/clang/19/include -internal-externc-isystem /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include -std=gnu++17 -fdeprecated-macro -ferror-limit 19 -stack-protector 1 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -fcxx-exceptions -fexceptions -fmax-type-align=16 -fcolor-diagnostics -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /var/folders/0w/4z5l9vds32nbkz7l22n8j6s80000gn/T/test_clang-047970.o -x c++ test_clang.cpp
clang -cc1 version 19.1.4 based upon LLVM 19.1.4 default target arm64-apple-darwin23.6.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /opt/homebrew/Cellar/llvm/19.1.4/bin/../include/c++/v1
 /opt/homebrew/Cellar/llvm/19.1.4/lib/clang/19/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks (framework directory)
End of search list.
 "/usr/bin/ld" -demangle -lto_library /opt/homebrew/Cellar/llvm/19.1.4/lib/libLTO.dylib -no_deduplicate -dynamic -arch arm64 -platform_version macos 14.0.0 14.0.0 -syslibroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -mllvm -enable-linkonceodr-outlining -o a.out /var/folders/0w/4z5l9vds32nbkz7l22n8j6s80000gn/T/test_clang-047970.o -lc++ -lSystem /opt/homebrew/Cellar/llvm/19.1.4/lib/clang/19/lib/darwin/libclang_rt.osx.a
Undefined symbols for architecture arm64:
  "typeinfo for __int128", referenced from:
      _main in test_clang-047970.o
ld: symbol(s) not found for architecture arm64
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

There is a similar problem for Apple Clang: https://developer.apple.com/forums/thread/679290

**Comment by jroelofs** [comment]
Seems plausible that this is the same class of problem as: https://github.com/llvm/llvm-project/issues/94677

**Comment by ArseniyKholod** [comment]
@eschnett found the workaround by using an additional linker flag: `-lc++abi`.

**Comment by jroelofs** [comment]
I missed that you’re using the clang driver to compile c++ files… you should use the clang++ symlink instead.

**Comment by philnik777** [comment]
CC @ldionne do we possibly not properly re-export the typeinfo for that?

**Comment by ArseniyKholod** [comment]
As may be a helpful note, the workaround by linking `-lc++abi` works somehow only for Apple aarch64, but not Apple x86_64.

**Comment by ldionne** [comment]
We do re-export `__ZTIn` (`typeinfo for __int128`) from here:

https://github.com/llvm/llvm-project/blob/d866005f6928a2a97e67866bedb26139d8cc27d9/libcxxabi/lib/fundamental-types.exp#L19

However, I can reproduce like this after installing with `brew install llvm`:

```
cat <<EOF | xcrun /opt/homebrew/opt/llvm/bin/clang++ -isysroot $(xcrun --show-sdk-path) -xc++ - -stdlib=libc++ -std=gnu++17 -frtti -v
#include <iostream>
#include <string>
#include <typeinfo>

int main()
{
  __int128 t = 1234567890;

  std::cout << typeid(__int128).name();

  return 0;
}
EOF
```

**Comment by ldionne** [comment]
After investigating, I think this is a duplicate of https://github.com/llvm/llvm-project/issues/79008. Your SDK is `MacOSX14.sdk` and that version does not contain the fix for #79008. The problem should be resolved if you use the macOS 15.4 SDK.
